File: dartsgame/dartsgame.go
package dartsgame

import (
	"fmt"
	"math"
)

type Darts struct {
	Height float64
	Speed  float64
	Angle  float64
}

type Target struct {
	Distance float64
	Height   float64
	Diameter float64
}

type Game struct {
	Darts  Darts
	Target Target
}

const gravity = 9.8

func (g *Game) Solve() (float64, error) {
	radian := 2.0 * math.Pi * g.Darts.Angle / 360.0
	arrive := g.Darts.Height + g.Target.Distance*math.Tan(radian) - gravity*g.Target.Distance*g.Target.Distance/(2.0*g.Darts.Speed*g.Darts.Speed*math.Cos(radian)*math.Cos(radian))
	diff := math.Abs(arrive - g.Target.Height)
	if diff < g.Target.Diameter/2.0 {
		return diff, nil
	} else {
		return 0, fmt.Errorf("miss")
	}
}

File: src/darts_game.rs
struct Darts {
    height: f64,
    speed: f64,
    angle: f64,
}

impl Darts {
    fn new(height: f64, speed: f64, angle: f64) -> Darts {
        Darts {
            height,
            speed,
            angle,
        }
    }
}

struct Target {
    distance: f64,
    height: f64,
    diameter: f64,
}

impl Target {
    fn new(distance: f64, height: f64, diameter: f64) -> Target {
        Target {
            distance,
            height,
            diameter,
        }
    }
}

struct Game {
    darts: Darts,
    target: Target,
}

const gravity: f64 = 9.8;

impl Game {
    fn new(darts: Darts, target: Target) -> Game {
        Game { darts, target }
    }

    fn solve(&self) -> Option<f64> {
        let arrive = self.darts.height + self.target.distance * tan(self.darts.angle) - (gravity * self.target.distance * self.target.distance / (2.0 * self.darts.speed * self.darts.speed * cos(self.darts.angle) * cos(self.darts.angle)));
        let diff = (arrive - self.target.height).abs();
        if diff < self.target.diameter / 2.0 {
            Some(diff)
        } else {
            None
        }
    }
}

fn cos(angle: f64) -> f64 {
    let rad = radian(angle);
    rad.cos()
}

fn tan(angle: f64) -> f64 {
    let rad = radian(angle);
    rad.tan()
}

fn radian(angle: f64) -> f64 {
    2.0 * std::f64::consts::PI * angle / 360.0
}

#[cfg(test)]
mod tests {
    use super::{Darts, Game, Target};

    #[test]
    fn test0() {
        let darts = Darts::new(10.0, 10.0, 10.0);
        let target = Target::new(10.0, 10.0, 10.0);
        let game = Game::new(darts, target);
        let want = 3.3;
        let got = game.solve().unwrap();
        let diff = (want-got).abs();
        assert!(diff < 0.1);
    }

    #[test]
    fn test1(){
        let darts = Darts::new(10.0, 15.0, 45.0);
        let target = Target::new(10.0, 10.0, 10.0);
        let game = Game::new(darts, target);
        assert_eq!(None, game.solve());
    }
}
